optuna/storages/_cached_storage.py

```diff
@@ -196,9 +197,15 @@ def set_trial_state(self, trial_id: int, state: TrialState) -> bool:
             cached_trial = self._get_cached_trial(trial_id)
             if cached_trial is not None:
                 self._check_trial_is_updatable(cached_trial)
+                previous_state = copy.copy(cached_trial.state)
                 updates = self._get_updates(trial_id)
                 cached_trial.state = state
                 updates.state = state
+
+                if state == TrialState.RUNNING and previous_state == TrialState.WAITING:
+                    updates.datetime_start = datetime.datetime.now()
+                    cached_trial.datetime_start = datetime.datetime.now()
```

Comment:
With the current cached storage logic, I don't think we'll enter this code since a waiting trial isn't cached. It might be worth adding an assertion. What do you think?

Reply:
I see. Previously added this partially because of logic in `test_set_trial_state`. Now fixed with assertion and modified tests.

Reply:
Ah, I meant to assert that a cached trial is never in the waiting state. A user may try to set the trial to running twice, in which case it should be a NOOP so the current assertion doesn't really hold in that case. I can follow this up in a PR.